File: Chapter 05/Python/section_5_4_server_monitor/after/infrastructure/adapters/twilio_adapter.py
from ...core.ports.alert_port import AlertPort
from ...infrastructure.external_libs.fake_libs import TwilioClient

class TwilioAdapter(AlertPort):
    """THE ADAPTER (Production).
    
    Bridges the internal AlertPort to the external Twilio SMS API.
    This class is the 'Clarity Engineer' that translates domain 
    intent into infrastructure-specific SDK calls.
    """

    def __init__(self, api_key: str, target_phone_number: str):
        """Configures the adapter with environment-specific secrets.
        
        Args:
            api_key (str): The 'God Mode' secret used to auth with Twilio.
            target_phone_number (str): The recipient of the SMS alerts.
        """
        self.client = TwilioClient(api_key)
        self.target_phone_number = target_phone_number

    def send_alert(self, message: str) -> None:
        """Encapsulates the chaotic 3rd-party SDK call.
        
        Args:
            message (str): The message provided by the Core.
        """
        self.client.send_sms(self.target_phone_number, message)
        print(f"(PROD ADAPTER) SMS sent via Twilio: {message}")

File: Chapter 05/JS/section_5_4_server_monitor/after/infrastructure/adapters/twilioAdapter.js
const AlertPort = require('../../core/ports/alertPort');
const { TwilioClient } = require('../externalLibs/fakeLibs');

/**
 * THE ADAPTER (Production).
 * Bridges the domain's AlertPort to the external Twilio SMS SDK.
 * This class acts as the 'Clarity Engineer' between inside and outside.
 */
class TwilioAdapter extends AlertPort {
    /**
     * Initializes the adapter with infrastructure-specific configuration.
     * * @param apiKey The secret key used for authenticating with Twilio.
     * @param targetPhoneNumber The destination number for the SMS alert.
     */
    constructor(apiKey, targetPhoneNumber) {
        super();
        // Initializing the client once at the boundary for resource efficiency.
        this.client = new TwilioClient(apiKey);
        this.targetPhoneNumber = targetPhoneNumber;
    }

    /**
     * Implements the Port by mapping a domain message to an SDK call.
     * * @param message The alert text to be sent.
     */
    sendAlert(message) {
        this.client.sendSms(this.targetPhoneNumber, message);
        console.log(`(PROD ADAPTER) SMS sent via Twilio: ${message}`);
    }
}

module.exports = TwilioAdapter;